### PR TITLE
Will/appeals 29406

### PR DESCRIPTION
**Pull request by wmedders21**
Resolves [Create correspondence_intakes table](https://jira.devops.va.gov/browse/APPEALS-29406)

# Description
Please explain the changes you made here.

- I generated a migration for the table `correspondence_intakes`
- I updated the schema by running the migration
- I added the associated active record model `CorrespondenceIntake`
- I tested the `belongs_to` relationship with `Correspondence`

## Acceptance Criteria
- [x] Code compiles correctly

## Testing Plan

1. Go to [Jira Issue/Test Plan Link](https://jira.devops.va.gov/browse/JIRA-12345) or list them below

- [ ] For feature branches merging into master: Was this deployed to UAT?

# Frontend
## User Facing Changes
 - [ ] Screenshots of UI changes added to PR & Original Issue

 BEFORE|AFTER
 ---|---

## Storybook Story
*For Frontend (Presentation) Components*
* [ ] Add a [Storybook](https://github.com/department-of-veterans-affairs/caseflow/wiki/Documenting-React-Components-with-Storybook) file alongside the component file (e.g. create `MyComponent.stories.js` alongside `MyComponent.jsx`)
* [ ] Give it a title that reflects the component's location within the overall Caseflow hierarchy
* [ ] Write a separate story (within the same file) for each discrete variation of the component

# Backend
## Database Changes
*Only for Schema Changes*

* [x] Add typical timestamps (`created_at`, `updated_at`) for new tables
* [x] Update column comments; include a "PII" prefix to indicate definite or potential [PII data content](https://github.com/department-of-veterans-affairs/appeals-team/blob/master/caseflow-team/0-how-we-work/pii-handbook.md#what-is-pii)
* [x] Have your migration classes inherit from `Caseflow::Migration`, especially when adding indexes (use `add_safe_index`) (see [Writing DB migrations](https://github.com/department-of-veterans-affairs/caseflow/wiki/Writing-DB-migrations))
* [x] Verify that `migrate:rollback` works as desired ([`change` supported functions](https://edgeguides.rubyonrails.org/active_record_migrations.html#using-the-change-method))
* [x] Perform query profiling (eyeball Rails log, check bullet and fasterer output)
* [ ] For queries using raw sql was an explain plan run by System Team
* [ ] Add appropriate indexes (especially for foreign keys, polymorphic columns, unique constraints, and Rails scopes)
* [x] Run `make check-fks`; add any missing foreign keys or add to `config/initializers/immigrant.rb` (see [Record associations and Foreign Keys](https://github.com/department-of-veterans-affairs/caseflow/wiki/Record-associations-and-Foreign-Keys))
* [x] Add `belongs_to` for associations to enable the [schema diagrams](https://department-of-veterans-affairs.github.io/caseflow/task_trees/schema/schema_diagrams) to be automatically updated
* [ ] Document any non-obvious semantics or logic useful for interpreting database data at [Caseflow Data Model and Dictionary](https://github.com/department-of-veterans-affairs/caseflow/wiki/Caseflow-Data-Model-and-Dictionary)

## Integrations: Adding endpoints for external APIs
* [ ] Check that Caseflow's external API code for the endpoint matches the code in the relevant integration repo
  * [ ] Request: Service name, method name, input field names
  * [ ] Response: Check expected data structure
  * [ ] Check that calls are wrapped in MetricService record block
* [ ] Check that all configuration is coming from ENV variables
  * [ ] Listed all new ENV variables in description
  * [ ] Worked with or notified System Team that new ENV variables need to be set
* [ ] Update Fakes
* [ ] For feature branches: Was this tested in Caseflow UAT

# Best practices
## Code Documentation Updates
- [ ] Add or update code comments at the top of the class, module, and/or component.

### Test Coverage
Did you include any test coverage for your code? Check below:
- [x] RSpec
- [ ] Jest
- [ ] Other

### Code Climate
Your code does not add any new code climate offenses? If so why?
- [ ] No new code climate issues added

### Monitoring
- [ ] Are performance metrics (e.g., response time, throughput) being tracked?
- [ ] Are key application components monitored (e.g., database, cache, queues)?
- [ ] Is there a system in place for setting up alerts based on performance thresholds?

### Logging
- [ ] Are logs being produced at appropriate log levels (debug, info, warn, error, fatal)?
- [ ] Are logs structured (e.g., using log tags) for easier querying and analysis?
- [ ] Are sensitive data (e.g., passwords, tokens) redacted or omitted from logs?
- [ ] Is log retention and rotation configured correctly?
- [ ] Are logs being forwarded to a centralized logging system if needed?

### Auditing
- [ ] Are user actions being logged for audit purposes?
- [ ] Are changes to critical data being tracked ?
- [ ] Are logs being securely stored and protected from tampering or exposing protected data?

### Error Handling
- [ ] Are errors being caught and handled gracefully?
- [ ] Are appropriate error messages being displayed to users?
- [ ] Are critical errors being reported to an error tracking system (e.g., Sentry, ELK)?
- [ ] Are unhandled exceptions being caught at the application level ?

### Exception Handling
- [ ] Are custom exceptions defined and used where appropriate?
- [ ] Is exception handling consistent throughout the codebase?
- [ ] Are exceptions logged with relevant context and stack trace information?
- [ ] Are exceptions being grouped and categorized for easier analysis and resolution?

### CODE REVIEW

**Review by youfoundmanesh — APPROVED** [review]

File: db/migrate/20231020172025_create_correspondence_intakes.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class CreateCorrespondenceIntakes < ActiveRecord::Migration[5.2]
+  def change
+    create_table :correspondence_intakes do |t|
+      t.integer :current_step, comment: "Tracks users progress on intake workflow"
+      t.bigint :related_appeals, array: true, comment: "Will be used to populate correspondence_appeals db table"
+      t.jsonb :added_tasks, array: true, comment: "Each object in the array will contain all relevant information to create the specific Task for either the Correspondence or the Related Appea"
```
Comment (jimfoley25) [inline db/migrate/20231020172025_create_correspondence_intakes.rb:6]: missing an "l" on Appeal

**Review by jimfoley25 — APPROVED** [review]
Looks good

**Review by jimfoley25 — COMMENTED** [review]
We will need to add a migration that changes the indexes (user reference & correspondence reference) according to [Creating a Migration Docs](https://github.com/department-of-veterans-affairs/caseflow/wiki/Writing-DB-migrations#creating-a-migration). Something like
```
class AddSafeIndexesToCorrespondenceIntakes < Caseflow::Migration
  def change
    add_safe_index :user, [:user_id], name: "index_on_user_id"
    add_safe_index :correspondence, [:correspondence_id], name: "index_on_correspondence_id"
  end
end
```

File: db/migrate/20231025161314_remove_indexes_from_correspondence_intakes.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class RemoveIndexesFromCorrespondenceIntakes < ActiveRecord::Migration[5.2]
+  def change
+    remove_index :correspondence_intakes, column: [:user_id], name: "index_correspondence_intakes_on_user_id"
+    remove_index :correspondence_intakes, column: [:correspondence_id], name: "index_correspondence_intakes_on_correspondence_id"
+  end
+end
```
Comment (jimfoley25) [inline db/migrate/20231025161314_remove_indexes_from_correspondence_intakes.rb:6]: remove this migration

File: db/migrate/20231020172025_create_correspondence_intakes.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+      t.string :error_reason, comment: "Exception details of when correspondence intake failed due to error"
+      t.datetime :canceled_at, comment: "Timestamp of when user cancelled correspondence intake"
+      t.string :canceled_reason, comment: "Details of reason user cancelled correspondence intake"
+      t.timestamps
+
+      t.references :user, foreign_key: true, comment: "Foreign key on users table"
+      t.references :correspondence, foreign_key: true, comment: "Foreign key on correspondences table"
+    end
+  end
+end
```
Comment (jimfoley25) [inline db/migrate/20231020172025_create_correspondence_intakes.rb:17]: suggest the following changes
change t.references :user ....... to

`t.integer :user_id, comment: ....`

change t.references :correspondence ...... to 

`t.integer :correspondence_id, comment: ....`

**Review by jimfoley25 — DISMISSED** [review]

**Review by jimfoley25 — APPROVED** [review]

**Review by jimfoley25 — APPROVED** [review]

**Review by jimfoley25 — APPROVED** [review]

**Review by jimfoley25 — APPROVED** [review]

**Pull request merged**